crates/spk-storage/src/walker.rs

```diff
@@ -0,0 +1,1155 @@
+                        yield RepoWalkerItem::Version(package_version.clone());
+
+                        let ident = Arc::clone(&package_version.ident);
+
+                        let builds = self.get_matching_builds(repository_name, repo, &ident).await?;
+                        for build in builds.iter() {
+                            yield RepoWalkerItem::Build(build.clone());
+
+                            let build_ident = Arc::new(build.spec.ident().clone());
+                            let components = self.get_matching_components(repository_name, repo, build_ident).await?;
+                            for component in components.iter() {
+                                yield RepoWalkerItem::Component(component.clone());
+
+                                let mut at_least_one_file = false;
+                                let mut file_stream = self.file_stream(repo, component.clone());
+                                while let Some(file_item) = self.convert_file_error(file_stream.try_next().await)? {
+                                    yield RepoWalkerItem::File(file_item);
+                                    at_least_one_file = true;
+                                }
+
+                                if self.emit_end_of_markers && at_least_one_file {
+                                    // Report the end of this component (after all its files and if it had any files)
+                                    yield RepoWalkerItem::EndOfComponent(component.clone())
+                                }
+                            }
+
+                            if self.emit_end_of_markers && !components.is_empty() {
+                                // Report the end of this build (after all the components in it)
+                                yield RepoWalkerItem::EndOfBuild(build.clone())
+                            }
+                        }
+
+                        if self.emit_end_of_markers && !builds.is_empty() {
+                            // Report the end of this version (after all its builds)
+                            yield RepoWalkerItem::EndOfVersion(package_version.clone())
+                        }
+                    }
+
+                    if self.emit_end_of_markers && !versions.is_empty() {
+                        // Report the end of this package (after all its versions)
+                        yield RepoWalkerItem::EndOfPackage(package.clone())
+                    }
+                }
+
+                if self.emit_end_of_markers && !packages.is_empty() {
+                    // Report the end of this repo (after all the packages in it)
+                    yield RepoWalkerItem::EndOfRepo(WalkedRepo { name: repo_name })
+                }
+            }
+        })
+    }
+}
+
+/// A builder for constructing a RepoWalker from various settings.
+///
+/// A default RepoWalker can made with:
+///
+///   let repo_walker_builder = RepoWalkerBuilder::new(repos);
+///   let repo_walker = builder.build();
```

Comment:
```suggestion
/// ```no_run
/// let repo_walker_builder = RepoWalkerBuilder::new(repos);
/// let repo_walker = builder.build();
/// ``` 
```

Reply:
This needed a rebase to pick up the changes in main that it had to turn these into doctests.